onnx/defs/controlflow/defs.cc

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) Facebook Inc. and Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "onnx/defs/schema.h"
+using namespace onnx;
+
+using SupportType = onnx::OpSchema::SupportType;
+
+OPERATOR_SCHEMA(If)
+    .SetSupportLevel(SupportType::EXPERIMENTAL)
+    .SetDoc("If conditional")
+    .Input(0, "cond", "Condition for the if", "B")
+    .Output(
+        0,
+        "outputs",
+        "Values that are live-out to the enclosing scope.",
+        "V",
+        OpSchema::Variadic)
+    .Attr(
+        "then_branch",
+        "Graph to run if condition is true. Has N outputs: values you wish to "
```

Comment:
branch graph nodes will have their own attribute data? I personally think using function in designing control flow ops is better. In that way, branch is a function (graph template), with parent node attributes/inputs/outputs, a graph could be instantiated.

Reply:
I don't understand what you mean. Can you provide an example of what this would look like?

Reply:
Hi @linkerzhang 

Do you think you could reproduce the example provided in the model definition using a function rather than a graph attribute?

  graph predict-net {
    %a = Constant[value = <Scalar Tensor [3]>]()
    %b = Constant[value = <Scalar Tensor [6]>]()
    %keepgoing = Constant[value = <Scalar Tensor [1]>]()
    %max_trip_count = Constant[value = <Scalar Tensor [10]>]()
    %keepgoing_out, %b_out, %user_defined_vals = Loop[body = <graph body-net>](%max_trip_count, %keepgoing, %b)
    return
  }

  graph body-net (
    %i[INT32, scalar]
    %keepgoing[BOOL, scalar]
    %b[INT32, scalar]
  ) {
    %my_local = Add(%a, %b)
    %b_out = Sub(%a, %b)
    %keepgoing_out = Greater(%my_local, %b_out)
    %user_defined_vals = Add(%b, %b)
    return %keepgoing_out, %b_out, %user_defined_vals
  }

Reply:
Regarding if-then-else: Like Dmytro mentioned earlier, it would be good to write down the conditions (about how the then-graph and else-graph should match). May be I missed this explanation, but I couldn't find it. Clearly both must have the same number of outputs. But are the outputs matched by name? (So, both branches have the same output-names?) I remember an earlier question about whether the corresponding outputs from the two branches must have the same type as well, but I can't find an answer to this.

Reply:
>Clearly both must have the same number of outputs.
Yeah, that's mentioned in the description for the sub-graphs
>But are the outputs matched by name?
No, it's purely positional. For e.x., I have the If op with `outputs = ['foo', 'bar', 'baz']`. I can have the then_branch net have `outputs = ['x', 'y', 'z']`, and else_branch with `outputs = ['a', 'b', 'c']`. Note that all three of these outputs are in different scopes. The op outputs refer to values defined in the enclosing scope. The then_branch outputs define names within the then_branch graph's scope, and the else_branch likewise. Imagine the `If` op encapsulates phi-nodes that reconcile its outputs to a value from the taken branch, e.g. `foo = phi(x, a)` and so on.
> I remember an earlier question about whether the corresponding outputs from the two branches must have the same type as well, but I can't find an answer to this.
That sounds reasonable, but outside of MS I don't think we have any infrastructure to validate any such specification of types, so can I ask for this to be a contribution on your side?